crates/amaru-ledger/src/rules/transaction/inputs.rs

```diff
@@ -52,16 +54,24 @@ where
         return Err(InvalidInputs::EmptyInputSet);
     }
 
-    // Check for disjoint reference inputs.
-    let intersection = match &reference_inputs {
-        Some(ref_inputs) => ref_inputs
-            .iter()
-            .filter(|ref_input| inputs.contains(ref_input))
-            .cloned()
-            .collect(),
-        None => Vec::new(),
-    };
+    let mut intersection = Vec::new();
 
+    if let Some(reference_inputs) = reference_inputs {
+        for reference_input in reference_inputs {
+            // Non-disjoint reference inputs
+            if inputs.contains(reference_input) {
+                intersection.push(reference_input.clone());
```

Comment:
It should actually be acceptable for this function to take ownership of the reference inputs and consume them here to avoid cloning. Maybe I am wrong but I can't think of another place where we might need those reference inputs?